### PR TITLE
ADD: RTL language support

### PR DESCRIPTION
**Pull request by marcosrdz**
Simulator Screen Shot - iPhone 12 Pro - 2021-03-10 at 21 56 58
Simulator Screen Shot - iPhone 12 Pro - 2021-03-10 at 21 57 00

Some components didn't auto update. Needs manual modifications

### CODE REVIEW

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/7pw3m5gx6qrgvdzyr4112xhy7w

[download apk](http://filestorage.bluewallet.io:1488/rtl-1615444293.apk)

**Comment by mcsaeid** [comment]
Great work. Here are some things I found:


1. Ellipsis (Settings)
1a. It would be better if placed on the right.



2. Creating a Wallet
2a. In the first screen, note that the new wallet, which has just been created, goes to the left, making it **completely invisible**. At first, I thought maybe I forgot to make it, but then I swiped to see it was there, only hidden. In other languages, this works because everything _is_ left to right.



3. GitHub Icon
3a. In English texts, as seen here, the icon must be placed on the left. (Note that the icons in the About section are rightly placed on the right because they are accompanied by Arabic/Persian texts.)





4. Single Arrows
4a. They must be flipped to face the other way.


5. Help Icon (Multi-signature)
5a. It would be better if placed on the right.

**Now here come the important ones.**


6. Seed Phrases
6a. Pretty straightforward. This may create confusion and has to be avoided. Seed phrases, being in English, must be displayed from left to right.




7. Bitcoin, Lightning, and Vault Icons
7a. Pretty straightforward. They need to be on the right side. (Note: If you look closely enough, the Lightning icon seems to be cut off a little from the top. I will probably create an issue for it.)

**Comment by mcsaeid** [comment]
Seven actually gave me the idea that icons need to be flipped. They should be kept on the left but flipped horizontally. I have edited the PNG files, but I don’t know how to upload these and make a pull request. Please advise.

Please take a look at the image below to see them all together.

vault-shape-ltr

Note that none of them have the same height. This creates the issue that I was talking about in the parentheses. If, however, they are the same height, I believe this will fix the problem.

**Review by jaonoctus — APPROVED** [review]

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/vmxv0yyent7nc70bq6bq0a3w2c

[download apk](http://filestorage.bluewallet.io:1488/rtl-1615919626.apk)

**Review by Overtorment — APPROVED** [review]
lgtm.
restart tests if necessary

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/cexzney7efd2e48j2xtt3vfcq0

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616033369.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/9926wg2b3df8qtyuymb2bjw3ug

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616076753.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/86tzmd8x1z8uky8u1mek5wu24w

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616117773.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/2fftw0qbuzpxht4ty5v0zc70gr

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616121854.apk)

**Comment by marcosrdz** [comment]
> Great work. Here are some things I found:
> 
> 
> 
>     1. Ellipsis (Settings)
>        1a. It would be better if placed on the right.
> 
> 
> 
> 
> 2. Creating a Wallet
> 2a. In the first screen, note that the new wallet, which has just been created, goes to the left, making it **completely invisible**. At first, I thought maybe I forgot to make it, but then I swiped to see it was there, only hidden. In other languages, this works because everything _is_ left to right.
> 
> 
> 
> 3. GitHub Icon
> 3a. In English texts, as seen here, the icon must be placed on the left. (Note that the icons in the About section are rightly placed on the right because they are accompanied by Arabic/Persian texts.)
> 
> 
> 
> 
> 
> 4. Single Arrows
> 4a. They must be flipped to face the other way.
> 
> 
> 5. Help Icon (Multi-signature)
> 5a. It would be better if placed on the right.
> 
> **Now here come the important ones.**
> 
> 
> 6. Seed Phrases
> 6a. Pretty straightforward. This may create confusion and has to be avoided. Seed phrases, being in English, must be displayed from left to right.
> 
> 
> 
> 
> 7. Bitcoin, Lightning, and Vault Icons
> 7a. Pretty straightforward. They need to be on the left side but they should be flipped. (Note: If you look closely enough, the Lightning icon seems to be cut off a little from the top. I will probably create an issue for it.)

I addressed the issues. The only one that hasnt been addressed is the wallets carousel on the main screen. That component lacks RTL support and would require more work. This will be eventually addressed.

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/x1fy0t1vt7ae6fkxz5thah8g1r

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616124260.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/vtbnjebd92ugm2c3fcrcx3bqh0

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616161869.apk)

**Comment by mcsaeid** [comment]
> I addressed the issues. The only one that hasnt been addressed is the wallets carousel on the main screen. That component lacks RTL support and would require more work. This will be eventually addressed.

Thank you, Marcos. I just checked. Very well done!

Here are some new things I found. At this point, you are probably like, “ah, man, here we go again.” 😂 Check these out.

2021-03-19 17-54-14 High Res Screenshot

When you create a new wallet, such as you see in the screenshot above, the opacity is not what it supposed to be, much like when a wallet is in the background and not selected. Look at the main version:

2021-03-19 17-58-27 High Res Screenshot

Here it seems that I am looking at the Bitcoin wallet, with the Lightning wallet being in the background with a lower opacity (as it should). That is not the case in the new RTL version.

2021-03-19 17-55-00 High Res Screenshot

Even when I create another wallet, they both look pale. Now when I go on the Lightning wallet, it looks fine:

2021-03-19 17-55-05 High Res Screenshot

Here is a multi-signature wallet:

2021-03-19 17-57-51 High Res Screenshot

2021-03-19 17-58-01 High Res Screenshot

The opacity seems to be correct only when the wallet is not selected/being looked at. For comparison, this is how it should be:

2021-03-19 17-58-34 High Res Screenshot

Also, one thing I have noticed is that the icons still look the same:

2021-03-19 17-55-13 High Res Screenshot

I appreciate the hard work. 🙏🏽

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/t2jevntkv0t8ym4p1qqp57v7dg

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616170188.apk)

**Comment by marcosrdz** [comment]
@mcsaeid  yea thats gonna require additional work since the carousel component does not have builtin RTL support. Recieve/Send icons  addressed.

**Comment by mcsaeid** [comment]
> 
> 
> @mcsaeid yea thats gonna require additional work since the carousel component does not have builtin RTL support. Recieve/Send icons addressed.

Understood.

I actually meant the PNG Bitcoin icon (as well as other icons) that needs flipping.

The send/receive icons were actually correct in that the arrow for the send button (placed on the left) pointed upward and the one for the receive button (placed on the right) pointed downward. I wonder if you flipped those.

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/y44gtvpdk5c4fn6da2e8t205n8

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616180026.apk)

**Comment by marcosrdz** [comment]
> > 
> 
> > 
> 
> > @mcsaeid yea thats gonna require additional work since the carousel component does not have builtin RTL support. Recieve/Send icons addressed.
> 
> 
> 
> Understood.
> 
> 
> 
> I actually meant the PNG Bitcoin icon (as well as other icons) that needs flipping.
@ncoelho 
> 
> 
> The send/receive icons were actually correct in that the arrow for the send button (placed on the left) pointed upward and the one for the receive button (placed on the right) pointed downward. I wonder if you flipped those.
> 
> 
> 
> Update: I just checked. Fantastic. They are absolutely correct.

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/bpezjbwvch8r81jb4y1xvcav1r

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616446468.apk)

**Comment by marcosrdz** [comment]
> > @mcsaeid yea thats gonna require additional work since the carousel component does not have builtin RTL support. Recieve/Send icons addressed.
> 
> Understood.
> 
> I actually meant the PNG Bitcoin icon (as well as other icons) that needs flipping.
> 
> The send/receive icons were actually correct in that the arrow for the send button (placed on the left) pointed upward and the one for the receive button (placed on the right) pointed downward. I wonder if you flipped those.
> 
> Update: I just checked. Fantastic. They are absolutely correct.

Wallet card images flipped.

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/9uz33mc6rr6qb7e0rw3b8zmmcw

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616541856.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/yvegkh2uhxe0re4mqw2ar4rae8

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616551370.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/2tdwxxt8u324g6hjncetvtt4v0

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616558026.apk)

**Comment by mcsaeid** [comment]
image

Shouldn’t the icons be 1) flipped (xxx-shape-rtl.png) 2) on the left side (as opposed to the main, LTR version, shown below)?

image

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/6cp7ebva649n3tp5jkgbjxq7bw

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616728225.apk)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/2ek51k2r26bdgy0rw6nvbu10w0

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616731000.apk)

**Comment by sha-265** [comment]
Maybe we should address an issue that forcing RTL layout, when using RTL devices:
Screenshot_2021-03-26 Appetize io - Run native mobile apps in your browser

I tried this potential fix here: https://github.com/sha-265/BlueWallet/commit/75a482e3aa3886671571167cd672c735435f7a7c but it didn't work (at least in emulator). The layout is still RTL when relaunching the app.

More info about this issue can be found [here](https://github.com/facebook/react-native/issues/10204). 

I'm not sure if this issue justify the effort, probably if you using RTL device, you will switch to RTL language in the app.

**Comment by marcosrdz** [comment]
> image
> 
> Shouldn’t the icons be 1) flipped (xxx-shape-rtl.png) 2) on the left side (as opposed to the main, LTR version, shown below)?
> 
> image

Fixed

**Comment by marcosrdz** [comment]
> Maybe we should address an issue that forcing RTL layout, when using RTL devices:
> Screenshot_2021-03-26 Appetize io - Run native mobile apps in your browser
> 
> I tried this potential fix here: [sha-265@75a482e](https://github.com/sha-265/BlueWallet/commit/75a482e3aa3886671571167cd672c735435f7a7c) but it didn't work (at least in emulator). The layout is still RTL when relaunching the app.
> 
> More info about this issue can be found [here](https://github.com/facebook/react-native/issues/10204) and [here](https://stackoverflow.com/questions/39815309/rtl-is-forced-in-rtl-devices).
> 
> But I'm not sure if this fixing this issue justify the effort, probably if you using RTL device, you will switch to RTL language in the app.

I replicated this bug. It has to be looked into as the suggested solution fails when doing a cold start.

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/90en773d9ax4uh1uex35racz04

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616847530.apk)

**Comment by mcsaeid** [comment]
> image
> 
> Shouldn’t the icons be 1) flipped (xxx-shape-rtl.png) 2) on the left side (as opposed to the main, LTR version, shown below)?
> 
> image

2021-03-27 17-09-00 High Res Screenshot
2021-03-27 17-09-13 High Res Screenshot

Thank you, @marcosrdz, for fixing the icon orientation. It looks fantastic now.

However, right now when you create your first wallet (or just a new one), the newly-created wallet goes **on the left** of the _Add Now_ card, making it not visible. The _Add Now_ card being on the right side of wallets also creates an issue in that in the RTL version, everything is read, viewed, and used from right to left. When you create a wallet, it would make sense to have the wallet show first (on the right side of the screen) and the _Add Now_ card to the left of it, and as you create or import new ones, they would stack up from right to left (based on the time of creation/import).

I raise this concern because if you look at the screenshots above (from the post I am quoting), you will see that the so-called wallet carousel was working perfectly correctly in that version.

Let me know what you think.

**Comment by marcosrdz** [comment]
@mcsaeid  made some changes. test it out and let me know.

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/fxuwx4cax11trjhvvj024xap8c

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616902042.apk)

**Comment by mcsaeid** [comment]
Much better!

2021-03-28 12-26-43 High Res Screenshot

However, as I mentioned in the previous comment, the _Add Now_ card must be placed on the left (far end) so that as wallets stack up, it doesn’t get in the way. What you see in the screenshot above is what the app defaults to when you first open it (or close and open it again). The first thing the user sees is not their wallet but the _Add Now_ card. They would have to swipe right to view their wallets.

2021-03-28 12-29-30 High Res Screenshot

One other thing that I have noticed is that the fiat balance in the RTL version, unlike BTC and sats, actually **goes to the left**. Take a look at what BTC looks like.

2021-03-28 12-29-37 High Res Screenshot

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/udy87j1kzkft5enkvtbe6p7krr

[download apk](http://filestorage.bluewallet.io:1488/rtl-1616956886.apk)

**Comment by marcosrdz** [comment]
@mcsaeid  please test out the latest changes

**Comment by mcsaeid** [comment]
> @mcsaeid please test out the latest changes

Super. Well done. The only issue that remains that I can think of is the fiat orientation in the wallets. Everything else looks nice and clean. I don’t know if this can be finalized now, but if so, I am sure if there are hidden issues, we will be able to find and address them later on as the new version comes out.

2021-03-29 15-27-44 High Res Screenshot

**Comment by marcosrdz** [comment]
> > @mcsaeid please test out the latest changes
> 
> Super. Well done. The only issue that remains that I can think of is the fiat orientation in the wallets. Everything else looks nice and clean. I don’t know if this can be finalized now, but if so, I am sure if there are hidden issues, we will be able to find and address them later on as the new version comes out.
> 
> 112828233-af004980-90a4-11eb-817c-8984e55344dd

Seems to be only affecting Android. Im unable to replicate on iOS.

**Comment by mcsaeid** [comment]
> Seems to be only affecting Android. Im unable to replicate on iOS.

I see. I guess we can live that.

**Comment by ncoelho** [comment]
Mind blowing work here. Congrats y'all

**Review by ncoelho — APPROVED** [review]

**Comment by mcsaeid** [comment]
Community-driven feedback from Twitter

**1. Pop-up window titles in RTL need to be right-aligned**
Here are some examples.


1a. Biometrics


1b. Wallet > First Receive > Backup Warning


1c. Wallet > First Receive > Notifications


1d. Settings > Security > Encrypted and Password Protected

**2. Settings with All English Items**
These two should be left-aligned.


2a. Settings > Currency


2b. Settings > Language

**3. Restart Alert in Specific Languages**

2021-04-01 15-52-14 High Res Screenshot
3a. _Most users_ who want to switch to another language, say to Farsi, cannot read or do not understand much English. In this case, the user was confused as to why the language settings hadn’t been applied, not knowing that they needed to _restart_ the app. They noted that the alert could be more straightforward. However, I believe it _is_, but, as I mentioned earlier, there will be users who just won’t be able to read the message and do what it says, thus creating confusion.

My suggestion is to show a specific alert for those few languages that do require restarting. I believe they are Arabic, Farsi, and Hebrew. If the user chooses Farsi, show the message in Farsi (have it translated). If the user chooses Arabic, an Arabic message will be shown. Showing a message in English might simply be ignored by those who do not know much English.

Credit goes to [So_Ghamari](https://twitter.com/So_Ghamari)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/xkrymm07ddhmf07e3k5kudqcnm

[download apk](http://filestorage.bluewallet.io:1488/rtl-1617339688.apk)

**Comment by marcosrdz** [comment]
@mcsaeid  do the alert title alignment issues occur if the device is configured to be RTL? Im wondering if these alerts follow system settings instead

**Comment by mcsaeid** [comment]
> @mcsaeid  do the alert title alignment issues occur if the device is configured to be RTL? Im wondering if these alerts follow system settings instead

Using Appetize, I never change the device settings, I simply open the app and start testing, and so the current issues happen with the device settings being its default _LTR_.

Now whether these _follow_ the system settings, that I am not sure, but I will test this by changing the device language to Farsi.

The alert titles, by the way, are not limited to the ones I have mentioned here. There were one or two more instances as well (that were found). I have assumed that all of them have the same issue and need to be addressed.

File: loc/index.js [hunk 1]
```diff
@@ -133,8 +134,25 @@ const setDateTimeLocale = async () => {
   }
   if (localeForDayJSAvailable) {
     dayjs.locale(lang.split('_')[0]);
+    const language = AvailableLanguages.find(language => language.value === lang.replace('_', '-'));
+    /* I18n Manager breaks testing. Mocking built-in RN modules is not so straightforward.
```
Comment (Overtorment) [inline loc/index.js:138]: hmm, we've got pretty good examples of mocks here tests/setup.js

File: loc/languages.js [hunk 2]
```diff
@@ -14,9 +14,9 @@ export const AvailableLanguages = Object.freeze([
...
   { label: 'Español (Latin America) (es_419)', value: 'es_419' },
   { label: 'Ελληνικά (EL)', value: 'el' },
-  { label: 'فارسی (FA)', value: 'fa' },
+  { label: 'فارسی (FA)', value: 'fa', isRTL: true },
   { label: 'Français (FR)', value: 'fr_fr' },
-  { label: 'עִברִית (HE)', value: 'he' },
+  { label: 'עִברִית (HE)', value: 'he', isRTL: true },
   { label: 'Italiano (IT)', value: 'it' },
   { label: 'Indonesia (ID)', value: 'id_id' },
   { label: 'Magyar (HU)', value: 'hu_hu' },
```
Comment (Overtorment) [inline loc/languages.js:22]: is japanese a RTL lang..?

**Review by Overtorment — APPROVED** [review]

File: loc/languages.js (hunk 2 above)
Comment (sha-265) [reply loc/languages.js:22]: @Overtorment do you mean Hebrew? If so, yes it is. If you mean Japanese, according to [Wikipedia](https://en.wikipedia.org/wiki/Right-to-left): 
> Arabic, Hebrew, Pashto, Persian, Urdu, and Sindhi are the most widespread RTL writing systems in modern times.
>
> Right-to-left can also refer to Text direction TDleft.svg top-to-bottom, right-to-left (TB-RL or TBRL) scripts such as Chinese, Japanese, and Korean, though in modern times they are also commonly written Text direction LTRdown.svg left to right.

Comment (Overtorment) [reply loc/languages.js:22]: got it. thanks. i just thought that maybe JP should be marked as RTL as well.

**Review by limpbrains — APPROVED** [review]
Great work!

**Comment by GladosBlueWallet** [comment]
Unbelievable. You, [subject name here], must be the pride of [subject hometown here]!

**Comment by EZ0010** [comment]
there is still an issue with RTL on the main screen for wallets on iOS
IMG_8556

**Comment by mcsaeid** [comment]
I can confirm this on iOS. Thank you, @EZ0010, for pointing it out. Here is what I have found:

image

The wallet that you see here, “Practice Area,” sits at the end of my lists of wallets in the the default orientation (English). In the RTL setting, however, it seems that the order of wallets is also flipped, which should not be. Take a look at the screen below:

image

As I swipe to the left, I am seeing my wallets in the reverse order. Now back to the English, LTR version, my first wallet on the list is a Lightning one:

image

image

The order, I believe, needs to be fixed.

**Comment by EZ0010** [comment]
I think some of the fixed done here worked perfectly on Android but not iOS for  some reason (I don't know why im not a developer).

the most needed fix right now is to fixed the main page where wallets are listed on above, other places are not that much needed to be fixed right now.

1-
Screen Shot 2021-05-07 at 4 48 03 AM

2-
Screen Shot 2021-05-07 at 4 49 12 AM

3-
Screen Shot 2021-05-07 at 4 50 20 AM

**Pull request merged**